### PR TITLE
Implement circuit breaker skip on 9628

### LINKED ISSUE #9628: FT 2.1: Implement CircuitBreaker skipOn
Implement the new spec requirements added by eclipse/microprofile-fault-tolerance#448

### PR DESCRIPTION
**Pull request by Joseph-Cass**
Fixes #9628

### CODE REVIEW

**Comment by Joseph-Cass** [comment]
#build

**Comment by Joseph-Cass** [comment]
#build

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/CircuitBreakerStateImplTest.java [hunk 1]
```diff
@@ -108,6 +108,91 @@ public void testFailOn() throws InterruptedException {
...
+
+        // 4x TestExceptionB -> 0% failure rate -> circuit closed
+        // Since TestExceptionB (or any of its superclasses) isn't mentioned in either failOn nor skipOn
+        runFailures(state, 4, TestExceptionB.class);
+        // assert circuit closed
+        runSuccesses(state, 1);
+    }
+
+    @Test
+    public void testMultipleExceptions() throws InterruptedException {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/CircuitBreakerStateImplTest.java]: It doesn't need any changes, but it's worth noting that this test includes an important case which wasn't in `testSkipOn`.

The important case is `failOn=TestExceptionBextended, skipOn=TestExceptionB` which asserts that `skipOn` always takes priority over `failOn`, even if an exception in `skipOn` is a superclass of an exception in `failOn`.

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/CircuitBreakerStateImplTest.java [hunk 2]
```diff
@@ -108,6 +108,91 @@ public void testFailOn() throws InterruptedException {
         runSuccesses(state, 1);
     }
 
+    @Test
+    public void testSkipOn() throws InterruptedException {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/CircuitBreakerStateImplTest.java]: Please update the copyright in this file.

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceCDIExtension.java [hunk 3]
```diff
@@ -113,7 +113,7 @@ public void beforeBeanDiscovery(@Observes BeforeBeanDiscovery beforeBeanDiscover
                         TimeoutConfig timeout = new TimeoutConfig(clazz, (Timeout) annotation);
                         timeout.validate();
                     } else if (annotation.annotationType() == CircuitBreaker.class) {
-                        CircuitBreakerConfig circuitBreaker = new CircuitBreakerConfig(clazz, (CircuitBreaker) annotation);
+                        CircuitBreakerConfig circuitBreaker = annotationConfigFactory.createCircuitBreakerConfig(clazz, (CircuitBreaker) annotation);
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceCDIExtension.java]: Update copyright in this file.

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceInterceptor.java [hunk 4]
```diff
@@ -125,7 +125,7 @@ private AggregatedFTPolicy processPolicies(InvocationContext context, BeanManage
                     retry = new RetryConfig(targetClass, (Retry) annotation);
                     retry.validate();
                 } else if (annotation.annotationType().equals(CircuitBreaker.class)) {
-                    circuitBreaker = new CircuitBreakerConfig(targetClass, (CircuitBreaker) annotation);
+                    circuitBreaker = annotationConfigFactory.createCircuitBreakerConfig(targetClass, (CircuitBreaker) annotation);
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceInterceptor.java]: Update copyright in this file.

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/AnnotationConfigFactory.java [hunk 5]
```diff
@@ -26,4 +27,7 @@
 
     public AsynchronousConfig createAsynchronousConfig(Class<?> annotatedClass, Asynchronous annotation);
 
+    public CircuitBreakerConfig createCircuitBreakerConfig(Method annotatedMethod, Class<?> annotatedClass, CircuitBreaker annotation);
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/AnnotationConfigFactory.java]: Update copyright in this file.

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/AnnotationConfigFactoryImpl.java [hunk 6]
```diff
@@ -36,4 +38,16 @@ public AsynchronousConfig createAsynchronousConfig(Class<?> annotatedClass, Asyn
         return new AsynchronousConfigImpl(annotatedClass, annotation);
     }
 
+    /** {@inheritDoc} */
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/AnnotationConfigFactoryImpl.java]: Update copyright in this file

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/AnnotationConfigFactoryImpl20.java [hunk 7]
```diff
@@ -36,4 +38,16 @@ public AsynchronousConfig createAsynchronousConfig(Class<?> annotatedClass, Asyn
         return new AsynchronousConfigImpl20(annotatedClass, annotation);
     }
 
+    /** {@inheritDoc} */
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/AnnotationConfigFactoryImpl20.java]: Update copyright in this file

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/CircuitBreakerConfigImpl.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import org.eclipse.microprofile.faulttolerance.CircuitBreaker;
+import org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.microprofile.faulttolerance.cdi.config.CircuitBreakerConfig;
+import com.ibm.ws.microprofile.faulttolerance.spi.CircuitBreakerPolicy;
+import com.ibm.ws.microprofile.faulttolerance.spi.FaultToleranceProvider;
+
+public class CircuitBreakerConfigImpl extends AbstractAnnotationConfig<CircuitBreaker> implements CircuitBreakerConfig {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/CircuitBreakerConfigImpl.java]: Update copyright in this file

File: dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/AnnotationConfigFactoryImpl21.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2019 IBM Corporation and others.
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/AnnotationConfigFactoryImpl21.java]: Copyright date on this file should be 2019

File: dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/test/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/CircuitBreakerConfig21Test.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    @After
+    public void after() {
+        ((Config13ProviderResolverImpl) ConfigProviderResolver.instance()).shutdown();
+        ConfigProviderResolver.setInstance(null);
+        FaultToleranceProviderResolver.setInstance(null);
+    }
+
+    @Test
+    public void defaultCircuitBreakerPolicyTest() throws NoSuchMethodException, SecurityException {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/test/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/CircuitBreakerConfig21Test.java]: Not sure about this test name.

It reads like you're testing the `CircuitBreakerPolicy`, but this is a test of the `CircuitBreakerConfig`

Also, you'd usually have `test` first.

I might change this to `testDefaultValues`, or `testCbConfigDefaultValues`

Similarly for `customCircuitBreakerPolicyTest`.

File: dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/test/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/CircuitBreakerConfig21Test.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        assertNotNull("FailOn not initialised to a default value ", testCBPolicy.getFailOn());
+
+        // Test a v2 Circuit Breaker Policy has a value
+        assertNotNull("SkipOn not initialised to a default value ", testCBPolicy.getSkipOn());
+
+        // Test a v1 Circuit Breaker Policy has the correct default value
+        assertArrayEquals("FailOn not initialised to the correct default value ", testCBPolicy.getFailOn(), new Class[] { Throwable.class });
+
+        // Test a v2 Circuit Breaker Policy has the correct default value
+        assertArrayEquals("SkipOn not initialised to the correct default value ", testCBPolicy.getSkipOn(), new Class[] {});
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/test/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/CircuitBreakerConfig21Test.java]: Not sure what the comments mean here. We don't have a v1 and v2 policy.

File: dev/com.ibm.ws.microprofile.faulttolerance.2.1_fat_tck/bnd.bnd [hunk 12]
```diff
@@ -19,8 +19,6 @@ src: \
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.1_fat_tck/bnd.bnd]: This commit is really a fixup for " Added code to run tck tests for fault tolerance 2.1. " - can you merge the two together.

**Comment by Azquelt** [comment]
Last build failed due to a failure to upload files to the file server.

#build

**Comment by Joseph-Cass** [comment]
#build

**Comment by Joseph-Cass** [comment]
#build

**Comment by Joseph-Cass** [comment]
#run-libby-bot

**Pull request merged**